pkg/bosh/manifest/manifest.go

```diff
@@ -1,12 +1,202 @@
+	Provides   map[string]interface{} `yaml:"provides,omitempty"`
+	Properties map[string]interface{} `yaml:"properties,omitempty"`
+}
+
+// VMResource from BOSH deployment manifest
+type VMResource struct {
+	CPU               int `yaml:"cpu"`
+	RAM               int `yaml:"ram"`
+	EphemeralDiskSize int `yaml:"ephemeral_disk_size"`
+}
+
+// Network from BOSH deployment manifest
+type Network struct {
+	Name      string   `yaml:"name"`
+	StaticIps []string `yaml:"static_ips,omitempty"`
+	Default   []string `yaml:"default,omitempty"`
+}
+
+// Update from BOSH deployment manifest
+type Update struct {
+	Canaries        int     `yaml:"canaries"`
+	MaxInFlight     string  `yaml:"max_in_flight"`
+	CanaryWatchTime string  `yaml:"canary_watch_time"`
+	UpdateWatchTime string  `yaml:"update_watch_time"`
+	Serial          bool    `yaml:"serial,omitempty"`
+	VMStratergy     *string `yaml:"vm_stratergy,omitempty"`
+}
+
+// MigratedFrom from BOSH deployment manifest
+type MigratedFrom struct {
+	Name string `yaml:"name"`
+	Az   string `yaml:"az,omitempty"`
+}
+
+// IPv6 from BOSH deployment manifest
+type IPv6 struct {
+	Enable bool `yaml:"enable"`
+}
+
+// JobDir from BOSH deployment manifest
+type JobDir struct {
+	Tmpfs     *bool  `yaml:"tmpfs,omitempty"`
+	TmpfsSize string `yaml:"tmpfs_size,omitempty"`
+}
+
+// Agent from BOSH deployment manifest
+type Agent struct {
+	Settings string `yaml:"settings,omitempty"`
+	Tmpfs    *bool  `yaml:"tmpfs,omitempty"`
+}
+
+// AgentEnvBoshConfig from BOSH deployment manifest
+type AgentEnvBoshConfig struct {
+	Password              string  `yaml:"password,omitempty"`
+	KeepRootPassword      string  `yaml:"keep_root_password,omitempty"`
+	RemoveDevTools        *bool   `yaml:"remove_dev_tools,omitempty"`
+	RemoveStaticLibraries *bool   `yaml:"remove_static_libraries,omitempty"`
+	SwapSize              *int    `yaml:"swap_size,omitempty"`
+	IPv6                  IPv6    `yaml:"ipv6,omitempty"`
+	JobDir                *JobDir `yaml:"job_dir,omitempty"`
+	Agent                 *Agent  `yaml:"agent,omitempty"`
+}
+
+// AgentEnv from BOSH deployment manifest
+type AgentEnv struct {
+	PersistentDiskFS           string              `yaml:"persistent_disk_fs,omitempty"`
+	PersistentDiskMountOptions []string            `yaml:"persistent_disk_mount_options,omitempty"`
+	AgentEnvBoshConfig         *AgentEnvBoshConfig `yaml:"bosh,omitempty"`
+}
+
 // InstanceGroup from BOSH deployment manifest
 type InstanceGroup struct {
-	Name      string
-	Instances int
+	Name               string                 `yaml:"name"`
+	Instances          int                    `yaml:"instances"`
+	Azs                []string               `yaml:"azs"`
+	Jobs               []Job                  `yaml:"jobs"`
+	VMType             string                 `yaml:"vm_type,omitempty"`
+	VMExtensions       []string               `yaml:"vm_extensions,omitempty"`
+	VMResources        *VMResource            `yaml:"vm_resources"`
```

Comment:
I'd prefer to avoid  pointers to struct in the data structure. My reasoning is, that we want to avoid shared mutable state, so we don't have side effects. We also want to avoid pointer nil checks.

I don't think performance, memory consumption is  a problem, since we don't change the struct a lot and it's rather small. I wouldn't try to optimize that without benchmarking real world usage first, anyhow.

Reply:
Ah, I see, we can't have optional fields in our struct unless they are pointers?

Reply:
Yes. all fields pop up with default values if we don't use pointers.